walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,110 @@
 
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.stream.Stream;
+import java.util.stream.Collectors;
+import java.util.Iterator;
 
 public final class FindMeetingQuery {
+  private static final long MAX_DURATION = TimeRange.WHOLE_DAY.duration();
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long minDuration = request.getDuration();
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+    if (minDuration > MAX_DURATION) {
+      return new ArrayList<TimeRange>();
+    }
+
+    List<TimeRange> mandatoryEventTimes = getBusyTimeRanges(events, mandatoryAttendees);
+    List<TimeRange> optionalEventTimes = getBusyTimeRanges(events, optionalAttendees);
+    List<TimeRange> allEventTimes =  new ArrayList<TimeRange>();
+    allEventTimes.addAll(mandatoryEventTimes);
+    allEventTimes.addAll(optionalEventTimes);
+
+    List<TimeRange> allAvailableTimes = getFreeTimeRanges(allEventTimes, minDuration);
+
+    // if there's meetings where all optional attendees can attend or if there's no mandatroy attendees
+    if (!allAvailableTimes.isEmpty() || mandatoryEventTimes.isEmpty()) {
+      return allAvailableTimes;
+    }
+
+    return getFreeTimeRanges(mandatoryEventTimes, minDuration);
+  }
+
+  private List<TimeRange> getFreeTimeRanges(List<TimeRange> busyTimes, long minDuration) {
+    busyTimes.sort(TimeRange.ORDER_BY_START);
+    removeNestedTimeRanges(busyTimes);
+    List<TimeRange> availableTimes = new ArrayList<TimeRange>();
+
+    TimeRange gap;
+    int start = TimeRange.START_OF_DAY;
+
+    if (busyTimes == null || busyTimes.isEmpty()) {
+      gap = TimeRange.WHOLE_DAY;
+      addIfValid(availableTimes, gap, minDuration);
+      return availableTimes;
+    }
+
+    for (int i = 0; i < busyTimes.size(); i++) {
+      gap = TimeRange.fromStartEnd(start, busyTimes.get(i).start(), false);
+      addIfValid(availableTimes, gap, minDuration);
+      start = busyTimes.get(i).end();
+    }
+
+    gap = TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, true);
+    addIfValid(availableTimes, gap, minDuration);
+
+    return availableTimes;
+  }
+
+  /**
+   * Checks if a gap is valid, and then add to availableTimes
+   */
+  private void addIfValid(List<TimeRange> availableTimes, TimeRange gap, long minDuration) {
+    if (isRangeValid(gap, minDuration)) {
+      availableTimes.add(gap);
+    }
+  }
+
+  private boolean isRangeValid(TimeRange timeRange, long minDuration) {
+    return (timeRange.start() < timeRange.end()) && (timeRange.duration() >= minDuration);
+  }
+
+  /**
+   * Removes any nested TimeRanges from the list given
+   */
+  private void removeNestedTimeRanges(List<TimeRange> timeRanges) {
```

Comment:
Is it enough to remove nested time ranges?
What about
```
|-----------|
          |-----------|
```

Reply:
It should be! The removeNestedTimeRanges() function only removes
`|-----------|
     |-----|`
In the case of overlapping time ranges, it should save the gap with an end time before the start time, which I filtered out via the isRangeValid() function